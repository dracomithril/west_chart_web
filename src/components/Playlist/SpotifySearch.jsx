import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpotify } from '@fortawesome/free-brands-svg-icons';
import PlaylistFormContainer from './PlaylistForm';
import RowSpotifySearch from './RowSpotifySearch';
import PlaylistInfo from './PlaylistInfo';
import actionTypes from '../../reducers/actionTypes';
import { createPlaylistAndAddTracks, searchForMusic } from '../../utils/spotify_utils';
import { chartObjectProps } from '../typeDefinitions';

export class SpotifySearch extends React.Component {
  onStartClickHandler = () => {
    const {
      selected, onStartClick, updateSingleSearch, spotifyUser,
    } = this.props;
    selected.forEach((elem) => {
      searchForMusic({ ...elem.search, id: elem.id }, spotifyUser.access_token).then(res => res
          && updateSingleSearch({
            field: 'items',
            value: res.value,
            id: res.id,
          }));
    });
    onStartClick && onStartClick();
  };

  handleCratePlaylistClick = ({ playlistName, isPrivate }) => {
    const { selected, spotifyUser, updatePlaylistInfo } = this.props;
    const selectedElements = selected
      .map(({ search }) => (search.selected !== undefined ? search.selected.uri : undefined))
      .filter(elem => elem !== undefined);
    createPlaylistAndAddTracks(
      spotifyUser.access_token,
      spotifyUser.id,
      playlistName,
      isPrivate,
      selectedElements,
    ).then(updatePlaylistInfo);
  };

  handleSearchClick = (search, id) => () => {
    const { spotifyUser, updateSingleSearch } = this.props;
    searchForMusic({ ...search, id }, spotifyUser.access_token).then(({ id: searchId, value }) => {
      updateSingleSearch({ id: searchId, value, field: 'items' });
    });
  };

  render() {
    const {
      selected, spotifyPlaylistInfo, spotifyUser, updateSingleSearch, onSwap, onClearHandler,
    } = this.props;

    const searchListView = selected.map(({ search = {}, id }) => (
      <RowSpotifySearch
        items={search.items}
        selected={search.selected}
        id={id}
        artist={search.artist}
        full_title={search.full_title}
        title={search.title}
        key={id}
        onSwap={onSwap}
        onUpdateClick={updateSingleSearch}
        onSearchClick={this.handleSearchClick(search, id)}
        onClearClick={onClearHandler}
      />
    ));

    return (
      <div className="spotify-search">
        <div className="spotify-search__header">
          create
          <FontAwesomeIcon icon={faSpotify} />
          spotify playlist:
        </div>
        {spotifyPlaylistInfo.url !== null && <PlaylistInfo {...spotifyPlaylistInfo} />}

        <PlaylistFormContainer
          onCreatePlaylistClick={this.handleCratePlaylistClick}
          onStartClick={this.onStartClickHandler}
          hasElements={selected.length > 0}
          isUserLogged={Boolean(spotifyUser.id)}
        />
        <div className="spotify-search_list">
          {searchListView.length > 0 && searchListView}
        </div>
      </div>
    );
  }
}

SpotifySearch.contextTypes = {
  store: PropTypes.shape(),
};
SpotifySearch.propTypes = {
  selected: PropTypes.arrayOf(chartObjectProps),
  spotifyUser: PropTypes.shape(),
  spotifyPlaylistInfo: PropTypes.shape(),
  onStartClick: PropTypes.func,
  updateSingleSearch: PropTypes.func,
  updatePlaylistInfo: PropTypes.func,
  onSwap: PropTypes.func,
  onClearHandler: PropTypes.func,
};
SpotifySearch.defaultProps = {
  selected: [],
};

const mapStateToProps = ({ spotifyPlaylistInfo, spotifyUser } /* , ownProps */) => ({
  spotifyPlaylistInfo,
  spotifyUser,
});

const mapDispatchToProps = dispatch => ({
  updateSingleSearch: (target) => {
    dispatch({
      type: actionTypes.UPDATE_SINGLE_SEARCH,
      field: target.field,
      value: target.value,
      id: target.id,
    });
  },
  updatePlaylistInfo: (info) => {
    dispatch({
      type: actionTypes.UPDATE_PLAYLIST_INFO,
      value: info,
    });
  },
  onSwap: (search_id) => {
    dispatch({
      type: actionTypes.SWAP_FIELDS,
      id: search_id,
    });
  },
  onClearHandler: (target) => {
    dispatch({
      type: actionTypes.CLEAR_SELECTED,
      id: target.id,
    });
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(SpotifySearch);
