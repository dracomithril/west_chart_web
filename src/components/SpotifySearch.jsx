/**
 * Created by Gryzli on 09.04.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import PlaylistForm from './PlaylistForm';
import RowSpotifySearch from './RowSpotifySearch';
import PlaylistInfo from './PlaylistInfo';

export default class SpotifySearch extends React.Component {
  /* istanbul ignore next */
  componentDidMount() {
    console.info('component SpotifySearch did mount');
  }
  /* istanbul ignore next */
  componentWillUnmount() {
    console.info('component SpotifySearch unmounted');
  }

  render() {
    const { store } = this.context;
    const { search_list, sp_playlist_info } = store.getState();
    const search_list_view = search_list.map(search_elem => (
      <RowSpotifySearch search_elem={search_elem} key={search_elem.id} />
    ));
    return (
      <div>
        <h3 id="list">
          {'Create '}
          <i className="fa fa-spotify" aria-hidden="true">
            Spotify playlist:
          </i>
          <PlaylistInfo info={sp_playlist_info} />
          <PlaylistForm {...this.props} />
        </h3>
        {search_list_view.length > 0 && (
          <table style={{ marginLeft: 10 }}>
            <thead>
              <tr>
                <th>Artist & Title</th>
                <th>Options:</th>
                <th>Selected:</th>
              </tr>
            </thead>
            <tbody>{search_list_view}</tbody>
          </table>
        )}
      </div>
    );
  }
}
SpotifySearch.contextTypes = {
  store: PropTypes.object,
};
SpotifySearch.propTypes = {
  selected: PropTypes.array,
};
