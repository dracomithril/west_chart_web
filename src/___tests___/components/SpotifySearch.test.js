/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import SpotifySearch from './../../components/SpotifySearch';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../data/initial_state.json');

const mockStore = configureMockStore([]);
describe('<SpotifySearch/>', () => {
  it('renders without crashing', () => {
    const store = mockStore(initial_state);
    const wrapper = shallow(
      <SpotifySearch selected={[]}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();

  });
});
