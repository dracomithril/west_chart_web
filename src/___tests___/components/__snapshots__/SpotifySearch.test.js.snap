// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpotifySearch/> renders without crashing 1`] = `
<div>
  <h3
    id="list"
  >
    Create 
    <i
      aria-hidden="true"
      className="fa fa-spotify"
    >
      Spotify playlist:
    </i>
    <PlaylistInfo
      info={
        Object {
          "url": null,
        }
      }
    />
    <PlaylistForm
      selected={Array []}
    />
  </h3>
</div>
`;
